// ngx-widget-grid/dist/models/Grid.model.js
import { CellSize } from './CellSize.model';
var Grid = /** @class */ (function () {
    function Grid(rows, columns) {
        this._widgets = [];
        this._rows = 3;
        this._columns = 3;
        if (rows) {
            this._rows = rows;
        }
        if (columns) {
            this._columns = columns;
        }
        this._cellSize = new CellSize(this._rows, this._columns);
    }
    Object.defineProperty(Grid.prototype, "rows", {
        get: function () {
            return this._rows;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Grid.prototype, "columns", {
        get: function () {
            return this._columns;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Grid.prototype, "cellSize", {
        get: function () {
            return this._cellSize;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Grid.prototype, "widgets", {
        get: function () {
            return this._widgets;
        },
        enumerable: true,
        configurable: true
    });
    Grid.prototype.add = function (widget) {
        this._widgets.push(widget);
    };
    Grid.prototype.remove = function (widget) {
        var widgetIndex = this._widgets.indexOf(widget);
        if (widgetIndex > -1) {
            this._widgets.splice(widgetIndex, 1);
        }
    };
    Grid.prototype.removeAll = function () {
        this._widgets.splice(0);
    };
    Grid.prototype.hasWidget = function (widget) {
        var widgetIndex = this._widgets.indexOf(widget);
        return widgetIndex > -1;
    };
    Grid.prototype.resize = function (rows, columns) {
        columns = columns || 0;
        rows = rows || 0;
        if (columns > 0 && rows > 0 && columns !== this._columns || rows !== this._rows) {
            this._columns = columns;
            this._rows = rows;
            this._cellSize = new CellSize(this._rows, this._columns);
        }
    };
    return Grid;
}());
export { Grid };
