// ngx-widget-grid/dist/models/WidgetConfig.model.js
import { GridRectangle } from './GridRectangle.model';
import { Utils } from '../Utils';
var WidgetConfig = /** @class */ (function () {
    function WidgetConfig(gridArea) {
        this.top = 0;
        this.left = 0;
        this.width = 0;
        this.height = 0;
        this.id = this.generateUID();
        if (gridArea) {
            this.top = gridArea.top || 0;
            this.left = gridArea.left || 0;
            this.width = gridArea.width || 0;
            this.height = gridArea.height || 0;
        }
    }
    WidgetConfig.prototype.generateUID = function () {
        return 'vmDashboardWidget-' + ++WidgetConfig.widgetCount;
    };
    Object.defineProperty(WidgetConfig.prototype, "position", {
        get: function () {
            return new GridRectangle({ top: this.top, left: this.left, width: this.width, height: this.height });
        },
        set: function (gridArea) {
            this.top = Utils.isNumber(gridArea.top) ? gridArea.top : this.top;
            this.left = Utils.isNumber(gridArea.left) ? gridArea.left : this.left;
            this.width = Utils.isNumber(gridArea.width) ? gridArea.width : this.width;
            this.height = Utils.isNumber(gridArea.height) ? gridArea.height : this.height;
        },
        enumerable: true,
        configurable: true
    });
    WidgetConfig.prototype.getId = function () {
        return this.id;
    };
    WidgetConfig.widgetCount = 0;
    return WidgetConfig;
}());
export { WidgetConfig };
