// ngx-widget-grid/dist/ngx-widget-grid.module.js
import { NgModule } from '@angular/core';
import { NgxGridOverlayComponent } from './components/gridOverlay/gridOverlay.component';
import { NgxWidgetComponent } from './components/widget/widget.component';
import { NgxWidgetGridComponent } from './components/grid/grid.component';
import { CommonModule } from '@angular/common';
import { NgxWidgetMoverDirective } from './directives/widgetMover.directive';
import { NgxWidgetResizerDirective } from './directives/widgetResizer.directive';
var NgxWidgetGridModule = /** @class */ (function () {
    function NgxWidgetGridModule() {
    }
    NgxWidgetGridModule.forRoot = function () {
        return {
            ngModule: NgxWidgetGridModule,
            providers: []
        };
    };
    NgxWidgetGridModule.forChild = function () {
        return {
            ngModule: NgxWidgetGridModule,
            providers: []
        };
    };
    NgxWidgetGridModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: [
                        NgxGridOverlayComponent,
                        NgxWidgetComponent,
                        NgxWidgetGridComponent,
                        NgxWidgetMoverDirective,
                        NgxWidgetResizerDirective
                    ],
                    providers: [],
                    exports: [
                        NgxWidgetComponent,
                        NgxWidgetGridComponent
                    ]
                },] },
    ];
    /** @nocollapse */
    NgxWidgetGridModule.ctorParameters = function () { return []; };
    return NgxWidgetGridModule;
}());
export { NgxWidgetGridModule };
